### ISSUE TITLE
Upsampling the output from self.upscore

### ISSUE THREAD

**Issue by keyurparalkar** [wkentaro/pytorch-fcn#91]
Hello, I am new to this community and I have doubt regarding the [fcn32.py](https://github.com/wkentaro/pytorch-fcn/blob/master/torchfcn/models/fcn32s.py). It's getting hard  for me to understand that the output of ```self.upscore``` is ```[batch_size,21,64,64]```, then  how it doesn't generate any error while calculating the loss with the ground-truth. According to me, the output should be ```[batch_size,21,64,64]``` and then it should be rescaled to size ```[batch_size,21,224,224]``` If your input image is of ```224x224x3``` size. Am I right over here ? How is this upsampling taking place ? how does our ```self.upscore``` output size matches with ground-truth size ?

**Comment by wkentaro** [comment]
It is upsampled by the deconv layer.

**Comment by keyurparalkar** [comment]
Yes I know it is upsampled by deconv layer but dimensions of that particular predictions should match with input right ? When I tried to implement the exact architecture as given in fcn32.py I found that after using conv. Transpose 2d with stride 32 it gave me output as image with size ```batxh_size,num_classes,64,64```. How should get it equivalent to my input size ?

**Comment by wkentaro** [comment]
It should be equevalent to the input image.
The decov layer has stride 32 and it gives you output upsamled by 32 to be
224 in your case.
I think something wrong in your architecture.
Did you try the script I am hosting on GitHub?

**Comment by keyurparalkar** [comment]
This is my implementation Inspired by your script by using the bilinear kernel for convtranspose2d:
```
#Getting upsampling weights for 2d bilinear kernel:
def get_upsampling_weight(in_channels,out_channels,kernel_size):
    factor = (kernel_size - 1) // 2 
    if kernel_size % 2 == 1:
        center = factor -1
    else:
        center = factor - 0.5
        
    og = np.ogrid[:kernel_size,:kernel_size]
    filt = (1 - abs(og[0] - center) / factor) * (1 - abs(og[1] - center) / factor)
    weight = np.zeros((in_channels, out_channels, kernel_size, kernel_size),dtype=np.float64)
    weight[range(in_channels),range(out_channels),:,:] = filt
    return torch.from_numpy(weight).float()

#TODO:=====================================LOADING PRETRAINED RESNET50 MODEL======================================================

model1 = models.resnet50(pretrained=True)
# print(model1.children)

# only_feature_extractor_model = {k:v for k,v in model1.state_dict().items() if k.split('.')[0] == 'features'}
# print('ONLY features = \n',only_feature_extractor_model.keys())
#============================================================


# Freezing the layers except fc layers
cnt = 0
for elem in model1.children():
    cnt+=1
    if(cnt <=9):
        for param in elem.parameters():
            param.requires_grad = False

#Creating FCN custom model.
class FCN(nn.Module):
    def __init__(self):
        super(FCN,self).__init__()
        self.features = nn.Sequential(*list(model1.children())[:-1])
        self.classifier = nn.Sequential(
                                    nn.Conv2d(2048,2048,1),
                                    nn.ReLU(),
                                    nn.Conv2d(2048,21,1),                                    
                                    nn.ReLU(),
                                    nn.Conv2d(21,21,1)
        )
        self.upscore0 = nn.ConvTranspose2d(21,21,64,stride=32,bias=False)
        self.upsample = nn.UpsamplingBilinear2d(size=(224,224))
        self._initialize_weights()
        
    def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m,nn.ConvTranspose2d):
                assert m.kernel_size[0] == m.kernel_size[1]
                initial_weight = get_upsampling_weight(m.in_channels,m.out_channels,m.kernel_size[0])
                m.weight.data.copy_(initial_weight)
        
    def forward(self,x):
        h = self.features(x)
        h = self.classifier(h)
        h = self.upscore0(h)
        h = self.upsample(h)
        return h
    

#load the custom FCN model:
model2 = FCN()
# print(model2)
#load pretrained weights of VGG16 features only into FCN's features.
# model2.features.load_state_dict(only_feature_extractor_model)
# print(model2.features.state_dict().keys())

#For confirming that all the pretrained weights from Vgg16 are transfered to FCN custom model
# print(list(model2.features.parameters()) == list(model1.parameters())) # True
```

I am using ConvTranspose2d with stride =32 and kernel_size=64 but still I get output_size = to [batch_size,21,64,64]. So because of this I used UpsamplingBilinear2d to rescale it to 224x224. Any suggestions on what might be the cause of this ?

**Comment by keyurparalkar** [comment]
any suggestions?

**Comment by keyurparalkar** [comment]
I have managed to have equal Input and output sizes with resnet34 architecture by adding a padding of (100,100) on first conv. layer. I also got to know the importance of the line ```h = h[:, :, 19:19 + x.size()[2], 19:19 + x.size()[3]].contiguous()``` in your fcn32 architecture. It is basically use to clip the 21 channel output spatially. I also found out that the output shape from the model must be greater than input size so that it can be clipped to match the input size(spatially). This is my code for resnet34 architecture:
```
#TODO:=====================================LOADING PRETRAINED RESNET50 MODEL======================================================

model1 = models.resnet34(pretrained=True)
# print(model1.children)
# only_feature_extractor_model = {k:v for k,v in model1.state_dict().items() if k.split('.')[0] == 'features'}
# print('ONLY features = \n',only_feature_extractor_model.keys())
#============================================================


# Freezing the layers except fc layers
cnt = 0
for elem in model1.children():
    cnt+=1
    if(cnt <=8):
        for param in elem.parameters():
            param.requires_grad = False

#Creating FCN custom model.
class FCN(nn.Module):
    def __init__(self):
        super(FCN,self).__init__()
        self.conv0 = nn.Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(100,100), bias=False)
        self.features = nn.Sequential(*list(model1.children())[1:-2])
        self.classifier = nn.Sequential(
                                    nn.Conv2d(512,512,7),
                                    nn.ReLU(inplace=True),
                                    nn.Dropout2d(),
                                    nn.Conv2d(512,512,1),
                                    nn.ReLU(inplace=True),
                                    nn.Dropout2d(),
                                    nn.Conv2d(512,21,1)
        )
        self.upscore0 = nn.ConvTranspose2d(21,21,64,stride=32,bias=False)
        self._initialize_weights()
        
    def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m,nn.ConvTranspose2d):
                assert m.kernel_size[0] == m.kernel_size[1]
                initial_weight = get_upsampling_weight(m.in_channels,m.out_channels,m.kernel_size[0])
                m.weight.data.copy_(initial_weight)
        
    def forward(self,x):
        h = self.conv0(x)
        h = self.features(h)
        h = self.classifier(h)
        h = self.upscore0(h)
        
        #The line below helps to add offset or you can set resize the image again to it's original input size
        # Only expection is that the output generated from the above self.upscore0() must be
        # greater than input size.
        h = h[:, :, 19:19 + x.size()[2], 19:19 + x.size()[3]].contiguous()
#         h = self.upsample(h)
        return h
    

#load the custom FCN model:
model2 = FCN()
# print(model2.children)
#load pretrained weights of VGG16 features only into FCN's features.
# model2.features.load_state_dict(only_feature_extractor_model)
# print(model2.features.state_dict().keys())

#For confirming that all the pretrained weights from Vgg16 are transfered to FCN custom model
# print(list(model2.features.parameters()) == list(model1.parameters())) # True
````

For the time being I was using the fcn32.py architecture i.e. based on vgg16. But I am getting blank prediction for your fcn32.py script. I have changed the no. of input and out kernels i.e. feature maps to 512 because of memory constraints. Here is my modified code for fcn32.py based on VGG16:
```
#===================== Creating VGG16 from STRACH ================================
class FCN32s(nn.Module):

    pretrained_model = \
        osp.expanduser('~/data/models/pytorch/fcn32s_from_caffe.pth')

    @classmethod
    def download(cls):
        return fcn.data.cached_download(
            url='http://drive.google.com/uc?id=0B9P1L--7Wd2vM2oya3k0Zlgtekk',
            path=cls.pretrained_model,
            md5='8acf386d722dc3484625964cbe2aba49',
        )

    def __init__(self, n_class=21):
        super(FCN32s, self).__init__()
        # conv1
        self.conv1_1 = nn.Conv2d(3, 64, 3, padding=100)
        self.relu1_1 = nn.ReLU(inplace=True)
        self.conv1_2 = nn.Conv2d(64, 64, 3, padding=1)
        self.relu1_2 = nn.ReLU(inplace=True)
        self.pool1 = nn.MaxPool2d(2, stride=2, ceil_mode=True)  # 1/2

        # conv2
        self.conv2_1 = nn.Conv2d(64, 128, 3, padding=1)
        self.relu2_1 = nn.ReLU(inplace=True)
        self.conv2_2 = nn.Conv2d(128, 128, 3, padding=1)
        self.relu2_2 = nn.ReLU(inplace=True)
        self.pool2 = nn.MaxPool2d(2, stride=2, ceil_mode=True)  # 1/4

        # conv3
        self.conv3_1 = nn.Conv2d(128, 256, 3, padding=1)
        self.relu3_1 = nn.ReLU(inplace=True)
        self.conv3_2 = nn.Conv2d(256, 256, 3, padding=1)
        self.relu3_2 = nn.ReLU(inplace=True)
        self.conv3_3 = nn.Conv2d(256, 256, 3, padding=1)
        self.relu3_3 = nn.ReLU(inplace=True)
        self.pool3 = nn.MaxPool2d(2, stride=2, ceil_mode=True)  # 1/8

        # conv4
        self.conv4_1 = nn.Conv2d(256, 512, 3, padding=1)
        self.relu4_1 = nn.ReLU(inplace=True)
        self.conv4_2 = nn.Conv2d(512, 512, 3, padding=1)
        self.relu4_2 = nn.ReLU(inplace=True)
        self.conv4_3 = nn.Conv2d(512, 512, 3, padding=1)
        self.relu4_3 = nn.ReLU(inplace=True)
        self.pool4 = nn.MaxPool2d(2, stride=2, ceil_mode=True)  # 1/16

        # conv5
        self.conv5_1 = nn.Conv2d(512, 512, 3, padding=1)
        self.relu5_1 = nn.ReLU(inplace=True)
        self.conv5_2 = nn.Conv2d(512, 512, 3, padding=1)
        self.relu5_2 = nn.ReLU(inplace=True)
        self.conv5_3 = nn.Conv2d(512, 512, 3, padding=1)
        self.relu5_3 = nn.ReLU(inplace=True)
        self.pool5 = nn.MaxPool2d(2, stride=2, ceil_mode=True)  # 1/32

        # fc6
        self.fc6 = nn.Conv2d(512, 512, 7)
        self.relu6 = nn.ReLU(inplace=True)
        self.drop6 = nn.Dropout2d()

        # fc7
        self.fc7 = nn.Conv2d(512, 512, 1)
        self.relu7 = nn.ReLU(inplace=True)
        self.drop7 = nn.Dropout2d()

        self.score_fr = nn.Conv2d(512, n_class, 1)
        self.upscore = nn.ConvTranspose2d(n_class, n_class, 64, stride=32,
                                          bias=False)

        self._initialize_weights()

    def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                m.weight.data.zero_()
                if m.bias is not None:
                    m.bias.data.zero_()
            if isinstance(m, nn.ConvTranspose2d):
                assert m.kernel_size[0] == m.kernel_size[1]
                initial_weight = get_upsampling_weight(
                    m.in_channels, m.out_channels, m.kernel_size[0])
                m.weight.data.copy_(initial_weight)

    def forward(self, x):
        h = x
        h = self.relu1_1(self.conv1_1(h))
        h = self.relu1_2(self.conv1_2(h))
        h = self.pool1(h)

        h = self.relu2_1(self.conv2_1(h))
        h = self.relu2_2(self.conv2_2(h))
        h = self.pool2(h)

        h = self.relu3_1(self.conv3_1(h))
        h = self.relu3_2(self.conv3_2(h))
        h = self.relu3_3(self.conv3_3(h))
        h = self.pool3(h)

        h = self.relu4_1(self.conv4_1(h))
        h = self.relu4_2(self.conv4_2(h))
        h = self.relu4_3(self.conv4_3(h))
        h = self.pool4(h)

        h = self.relu5_1(self.conv5_1(h))
        h = self.relu5_2(self.conv5_2(h))
        h = self.relu5_3(self.conv5_3(h))
        h = self.pool5(h)

        h = self.relu6(self.fc6(h))
        h = self.drop6(h)

        h = self.relu7(self.fc7(h))
        h = self.drop7(h)

        h = self.score_fr(h)

        h = self.upscore(h)
        
        #The line below helps to add offset or you can set resize the image again to it's original input size
        # Only expection is that the output generated from the above self.upsocre() must be
        # greater than input size.
        h = h[:, :, 19:19 + x.size()[2], 19:19 + x.size()[3]].contiguous()

        return h

    def copy_params_from_vgg16(self, vgg16):
        features = [
            self.conv1_1, self.relu1_1,
            self.conv1_2, self.relu1_2,
            self.pool1,
            self.conv2_1, self.relu2_1,
            self.conv2_2, self.relu2_2,
            self.pool2,
            self.conv3_1, self.relu3_1,
            self.conv3_2, self.relu3_2,
            self.conv3_3, self.relu3_3,
            self.pool3,
            self.conv4_1, self.relu4_1,
            self.conv4_2, self.relu4_2,
            self.conv4_3, self.relu4_3,
            self.pool4,
            self.conv5_1, self.relu5_1,
            self.conv5_2, self.relu5_2,
            self.conv5_3, self.relu5_3,
            self.pool5,
        ]
        for l1, l2 in zip(vgg16.features, features):
            if isinstance(l1, nn.Conv2d) and isinstance(l2, nn.Conv2d):
                assert l1.weight.size() == l2.weight.size()
                assert l1.bias.size() == l2.bias.size()
                l2.weight.data = l1.weight.data
                l2.bias.data = l1.bias.data
        for i, name in zip([0, 3], ['fc6', 'fc7']):
            l1 = vgg16.classifier[i]
            l2 = getattr(self, name)
            l2.weight.data = l1.weight.data.view(l2.weight.size())
            l2.bias.data = l1.bias.data.view(l2.bias.size())
            
model2 = FCN32s()
```
Can you tell me what might be the cause of this blank predictions ?
![fcn32_doubt](https://user-images.githubusercontent.com/14138515/47267904-72d96e00-d567-11e8-94bd-238dda34824c.png)

**Comment by hubutui** [comment]
Hey, I come across the similar question. For a input image of size 335x500, [fcn32s.py#L143](https://github.com/wkentaro/pytorch-fcn/blob/master/torchfcn/models/fcn32s.py#L143) give an output of size 384x544, I couldn't understand line 144:
```python
h = h[:, :, 19:19 + x.size()[2], 19:19 + x.size()[3]].contiguous()
```
Is 19 a magic number?

**Comment by wkentaro** [comment]
@hubutui See here: https://github.com/wkentaro/pytorch-fcn/issues/51
